codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1047 @@
+                    readInProgress = false;
+                    unsafe().recvBufAllocHandle().readComplete();
+                    pipeline().fireChannelReadComplete();
+                }
+                return flushPending;
+            } finally {
+                inFireChannelReadComplete = false;
+                flushPending = false;
+            }
+        }
+
+        private final class Http2ChannelUnsafe implements Unsafe {
+            private final VoidChannelPromise unsafeVoidPromise =
+                    new VoidChannelPromise(DefaultHttp2StreamChannel.this, false);
+            @SuppressWarnings("deprecation")
+            private RecvByteBufAllocator.Handle recvHandle;
+            private boolean flush;
+            private boolean inClose;
+
+            @Override
+            public void connect(final SocketAddress remoteAddress,
+                                SocketAddress localAddress, final ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
+
+            @SuppressWarnings("deprecation")
+            @Override
+            public RecvByteBufAllocator.Handle recvBufAllocHandle() {
+                if (recvHandle == null) {
+                    recvHandle = config().getRecvByteBufAllocator().newHandle();
+                }
+                return recvHandle;
+            }
+
+            @Override
+            public SocketAddress localAddress() {
+                return parent().unsafe().localAddress();
+            }
+
+            @Override
+            public SocketAddress remoteAddress() {
+                return parent().unsafe().remoteAddress();
+            }
+
+            @Override
+            public void register(EventLoop eventLoop, ChannelPromise promise) {
+                if (registered) {
+                    throw new UnsupportedOperationException("Re-register is not supported");
+                }
+
+                registered = true;
+
+                // Add the handler to the pipeline now that we are registered.
+                pipeline().addLast(inboundStreamHandler);
+
+                promise.setSuccess();
+
+                pipeline().fireChannelRegistered();
+                if (isActive()) {
+                    pipeline().fireChannelActive();
+                }
+            }
+
+            @Override
+            public void bind(SocketAddress localAddress, ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
+
+            @Override
+            public void disconnect(ChannelPromise promise) {
+                close(promise);
+            }
+
+            @Override
+            public void close(ChannelPromise promise) {
+                if (closePromise.isDone()) {
+                    promise.setFailure(new ClosedChannelException());
+                    return;
+                }
+                if (inClose) {
+                    return;
```

Comment:
what about `promise`? do we need to save the "original promise" and complete `promise` when it completes?